### PR TITLE
More upload functions

### LINKED ISSUE #34: Upload functions for Sample, Proposal and Instrument
The intention is to extend pyscicat to cover all SciCat endpoints in the future.

### PR DESCRIPTION
**Pull request by eilmiv**
Closes #34

### CODE REVIEW

**Review by dylanmcreynolds — APPROVED** [review]
Thank you for the PR!

File: pyscicat/client.py [hunk 1]
```diff
@@ -413,6 +416,208 @@ def datasets_attachment_create(
...
+        """Updates an existing instrument.
+        Note that in SciCat admin rights are required to upload instruments.
+
+        Parameters
+        ----------
+        instrument : Instrument
+            Instrument to update
+
+        pid
+            pid (or unique identifier) of instrument being updated. By default, pid is taken from instrument parameter.
```
Comment (dylanmcreynolds) [inline pyscicat/client.py:527]: Oops. This line is too long for flake8. Can you add a new line?

Comment (eilmiv) [reply pyscicat/client.py]: Yes, I forgot to run flake8. Thank you for the quick reply.

File: pyscicat/client.py [hunk 2]
```diff
@@ -413,6 +416,209 @@ def datasets_attachment_create(
...
+        return self._call_endpoint(
+            cmd="post",
+            endpoint="Samples",
+            data=sample,
+            operation="samples_create",
+        ).get("sampleId")
+
+    upload_sample = samples_create
+
+    def update_sample(self, sample: Sample, sampleId: str = None) -> str:
```
Comment (nitrosx) [inline pyscicat/client.py:450]: Can we change the name to this function to **_samples_update_** to be consistent with naming convention used with other endpoints?

File: pyscicat/client.py [hunk 3]
```diff
@@ -413,6 +416,209 @@ def datasets_attachment_create(
...
+        return self._call_endpoint(
+            cmd="post",
+            endpoint="Instruments",
+            data=instrument,
+            operation="instruments_create",
+        ).get("pid")
+
+    upload_instrument = instruments_create
+
+    def update_instrument(self, instrument: Instrument, pid: str = None) -> str:
```
Comment (nitrosx) [inline pyscicat/client.py:517]: Same as my previous comment, can we change the name to _**instruments_update**_?

File: pyscicat/client.py [hunk 4]
```diff
@@ -413,6 +416,209 @@ def datasets_attachment_create(
...
+        return self._call_endpoint(
+            cmd="post",
+            endpoint="Proposals",
+            data=proposal,
+            operation="proposals_create",
+        ).get("proposalId")
+
+    upload_proposal = proposals_create
+
+    def update_proposal(self, proposal: Proposal, proposalId: str = None) -> str:
```
Comment (nitrosx) [inline pyscicat/client.py:586]: Same as before, can we rename it **_proposals_update_**?

**Review by nitrosx — CHANGES_REQUESTED** [review]

File: pyscicat/client.py (hunk 2 above)
Comment (eilmiv) [reply pyscicat/client.py]: Yes. I agree; this is better. (I was trying to be consistent with the update_dataset endpoint, as all the existing update functions that have a name starting with the endpoint are `_replace` functions, but from what I can tell the next backend no longer supports `replaceOrCreate` operations.)

File: pyscicat/client.py (hunk 3 above)
Comment (eilmiv) [reply pyscicat/client.py]: See answer to [previous comment](https://github.com/SciCatProject/pyscicat/pull/35#discussion_r1090287857)

File: pyscicat/client.py (hunk 4 above)
Comment (eilmiv) [reply pyscicat/client.py]: See answer to [previous comment](https://github.com/SciCatProject/pyscicat/pull/35#discussion_r1090287857)

File: pyscicat/client.py (hunk 2 above)
Comment (nitrosx) [reply pyscicat/client.py]: Update_dataset has been long overdue to be updated to the correct name, aka datasets_update.
Actually, would be so kind to update that one too?

Comment (eilmiv) [reply pyscicat/client.py]: I updated it and kept the old name, like for the existing renamed functions.

**Review by nitrosx — APPROVED** [review]

**Pull request merged**